Edit search documentation (#28737)
gatsbyjs/gatsby: docs/docs/how-to/adding-common-features/adding-search.md

One way of doing this is to use the `js-search` library:
* [Adding Search with JS Search](/docs/adding-search-with-js-search)

There are two Gatsby plugins that support this as well:
* [gatsby-plugin-elasticlunr-search](/plugins/@gatsby-contrib/gatsby-plugin-elasticlunr-search)
* [gatsby-plugin-local-search](/plugins/gatsby-plugin-local-search)

### Use an API-based search engine

Another option is to use an external search engine. This solution is much more scalable as visitors to your site don't have to download your entire search index (which becomes very large as your site grows) in order to search your site. The trade-off is you'll need to pay for hosting the search engine or pay for a commercial search service.

There are many options available, including both self-hosted and commercially hosted open source:

- [ElasticSearch](https://www.elastic.co/products/elasticsearch) — OSS and has commercial hosting available
- [Solr](http://lucene.apache.org/solr/) — OSS and has commercial hosting available
- [Algolia](https://www.algolia.com/) — Commercial

Of these, the most common solution is Algolia. The Gatsby docs include a guide to adding Algolia to your site:

> * [Adding Search with Algoolia](/docs/adding-search-with-algolia)

Reviewer: ```suggestion
* [Adding Search with Algolia](/docs/adding-search-with-algolia)
```

When using Algolia, they host the search index and search engine for you. Your search queries will be sent to their servers which will respond with any results. For UI components, Algolia provides a [React library](https://github.com/algolia/react-instantsearch) has helpful components.

If you're building a documentation website you can use [Algolia's DocSearch feature](https://community.algolia.com/docsearch/). It will automatically create a search index from the content of your pages.

> Elasticsearch also has several React component libraries for search e.g. https://github.com/appbaseio/reactivesearch

Reviewer: ```suggestion
Elasticsearch also has several React component libraries for search, such as [ReactiveSearch](https://github.com/appbaseio/reactivesearch)
```
